### PR TITLE
[SYCL][Bindless] Add interop memory mapping to USM.

### PR DESCRIPTION
**Pull request by przemektmalon**
This patch introduces `map_external_linear_memory` to enable mapping interop memory to linear USM, returning a `void *`.

The PI function `piextMemMapExternalLinearMemory` has been added to enable this in the PI layer.

A Vulkan test case has been added to test this functionality.

### CODE REVIEW

**Comment by przemektmalon** [comment]
Associated [UR PR](https://github.com/oneapi-src/unified-runtime/pull/1882)

**Review by Seanst98 — APPROVED** [review]

**Review by konradkusiak97 — APPROVED** [review]

File: sycl/test-e2e/bindless_images/vulkan_interop/sampled_images_USM.cpp [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+    q.submit([&](sycl::handler &cgh) {
+      auto outAcc = buf.template get_access<sycl::access_mode::write>(
+          cgh, outBufferRange);
+      cgh.parallel_for<KernelName>(
+          sycl::nd_range<NDims>{globalSize, localSize},
+          [=](sycl::nd_item<NDims> it) {
+            size_t dim0 = it.get_global_id(0);
+            size_t dim1 = it.get_global_id(1);
+
+            // Normalize coordinates -- +0.5 to look towards centre of pixel
```
Comment (cperkinsintel) [inline sycl/test-e2e/bindless_images/vulkan_interop/sampled_images_USM.cpp:100]: thanks for the comment!

**Review by cperkinsintel — APPROVED** [review]

**Comment by przemektmalon** [comment]
Friendly ping @intel/dpcpp-nativecpu-pi-reviewers

**Review by DBDuncan — APPROVED** [review]

**Review by uwedolinsky — APPROVED** [review]
NativeCPU 👍

File: sycl/cmake/modules/FetchUnifiedRuntime.cmake [hunk 2]
```diff
@@ -115,13 +115,13 @@ if(SYCL_PI_UR_USE_FETCH_CONTENT)
       CACHE PATH "Path to external '${name}' adapter source dir" FORCE)
   endfunction()
 
-  set(UNIFIED_RUNTIME_REPO "https://github.com/oneapi-src/unified-runtime.git")
+  set(UNIFIED_RUNTIME_REPO "https://github.com/przemektmalon/unified-runtime.git")
   # commit bc1a28ede0df7f837047b632e00437587672c134
   # Author: Omar Ahmed <omar.ahmed@codeplay.com>
   # Date:   Mon Jul 29 16:44:58 2024 +0100
   #     Merge pull request #1819 from DBDuncan/sean/rename-interop-to-external
   #    [Bindless][Exp] Rename interop related structs/funcs with "external"
```
Comment (omarahmed1111) [inline sycl/cmake/modules/FetchUnifiedRuntime.cmake:123]:
```suggestion
  set(UNIFIED_RUNTIME_REPO "https://github.com/oneapi-src/unified-runtime.git")
  # commit c5d2175b5823d5b74de1e7e0d6081ab6d885bc34
  # Merge: 99489ad4 c86beb60
  # Author: Omar Ahmed <omar.ahmed@codeplay.com>
  # Date:   Wed Jul 31 14:52:26 2024 +0100
  #     Merge pull request #1882 from przemektmalon/przemek/interop-map-memory
  #    [Bindless][Exp] Add interop memory mapping to USM.
```

File: sycl/cmake/modules/FetchUnifiedRuntime.cmake [hunk 3]
```diff
@@ -115,13 +115,13 @@ if(SYCL_PI_UR_USE_FETCH_CONTENT)
...
-  set(UNIFIED_RUNTIME_TAG bc1a28ede0df7f837047b632e00437587672c134)
+  set(UNIFIED_RUNTIME_TAG c86beb60ba9c210011635f030e418289f221e2cb)
```
Comment (omarahmed1111) [inline sycl/cmake/modules/FetchUnifiedRuntime.cmake:124]:
```suggestion
  set(UNIFIED_RUNTIME_TAG c5d2175b5823d5b74de1e7e0d6081ab6d885bc34)
```

File: sycl/cmake/modules/FetchUnifiedRuntime.cmake (hunk 2 above)
Comment (DBDuncan) [reply sycl/cmake/modules/FetchUnifiedRuntime.cmake]: Done.

File: sycl/cmake/modules/FetchUnifiedRuntime.cmake (hunk 3 above)
Comment (DBDuncan) [reply sycl/cmake/modules/FetchUnifiedRuntime.cmake]: Done.

**Review by pbalcer — APPROVED** [review]

**Comment by omarahmed1111** [comment]
Included PRs in this UR bump:
* https://github.com/oneapi-src/unified-runtime/pull/1737
* https://github.com/oneapi-src/unified-runtime/pull/1790
* https://github.com/oneapi-src/unified-runtime/pull/1399
* https://github.com/oneapi-src/unified-runtime/pull/1908
* https://github.com/oneapi-src/unified-runtime/pull/1884
* https://github.com/oneapi-src/unified-runtime/pull/1880
* https://github.com/oneapi-src/unified-runtime/pull/1882

**Comment by DBDuncan** [comment]
@intel/llvm-gatekeepers Should be ok to merge? Does not touch HIP. Bindless images is not supported on AMD yet.

**Comment by sarnex** [comment]
@intel/bindless-images-reviewers Could this postcommit fail be related?

```
FAIL: SYCL-Unit :: program_manager/arg_mask/./ProgramManagerArgMaskTests/0/1 (1104 of 1159)
******************** TEST 'SYCL-Unit :: program_manager/arg_mask/./ProgramManagerArgMaskTests/0/1' FAILED ********************
Script(shard):
--
GTEST_OUTPUT=json:/__w/llvm/llvm/build/tools/sycl/unittests/program_manager/arg_mask/./ProgramManagerArgMaskTests-SYCL-Unit-351801-0-1.json GTEST_SHUFFLE=0 GTEST_TOTAL_SHARDS=1 GTEST_SHARD_INDEX=0 /__w/llvm/llvm/build/tools/sycl/unittests/program_manager/arg_mask/./ProgramManagerArgMaskTests
--

[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from EliminatedArgMask
[ RUN      ] EliminatedArgMask.KernelBundleWith2Kernels
terminate called after throwing an instance of 'sycl::_V1::exception'
  what():  A local accessor must not be used in a SYCL kernel function that is invoked via single_task or via the simple form of parallel_for that takes a range parameter.
 #0 0x00007facd7034aa8 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/__w/llvm/llvm/build/./lib/libLLVMSupport.so.19.0git+0x182aa8)
 #1 0x00007facd703273c llvm::sys::RunSignalHandlers() (/__w/llvm/llvm/build/./lib/libLLVMSupport.so.19.0git+0x18073c)
 #2 0x00007facd7035158 SignalHandler(int) Signals.cpp:0:0
 #3 0x00007facd6990520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
 #4 0x00007facd69e49fc pthread_kill (/lib/x86_64-linux-gnu/libc.so.6+0x969fc)
 #5 0x00007facd6990476 gsignal (/lib/x86_64-linux-gnu/libc.so.6+0x42476)
 #6 0x00007facd69767f3 abort (/lib/x86_64-linux-gnu/libc.so.6+0x287f3)
 #7 0x00007facd6d20b9e (/lib/x86_64-linux-gnu/libstdc++.so.6+0xa2b9e)
 #8 0x00007facd6d2c20c (/lib/x86_64-linux-gnu/libstdc++.so.6+0xae20c)
 #9 0x00007facd6d2c277 (/lib/x86_64-linux-gnu/libstdc++.so.6+0xae277)
#10 0x00007facd6d2c4d8 (/lib/x86_64-linux-gnu/libstdc++.so.6+0xae4d8)
#11 0x0000000000493baf getKernelArgMaskFromBundle(sycl::_V1::kernel_bundle<(sycl::_V1::bundle_state)0> const&, std::shared_ptr<sycl::_V1::detail::queue_impl>) (/__w/llvm/llvm/build/tools/sycl/unittests/program_manager/arg_mask/./ProgramManagerArgMaskTests+0x493baf)
#12 0x0000000000495689 EliminatedArgMask_KernelBundleWith2Kernels_Test::TestBody() (/__w/llvm/llvm/build/tools/sycl/unittests/program_manager/arg_mask/./ProgramManagerArgMaskTests+0x495689)
#13 0x00007facd7109160 testing::Test::Run() (/__w/llvm/llvm/build/./lib/libllvm_gtest.so.19.0git+0x3e160)
#14 0x00007facd710a5c9 testing::TestInfo::Run() (/__w/llvm/llvm/build/./lib/libllvm_gtest.so.19.0git+0x3f5c9)
#15 0x00007facd710b335 testing::TestSuite::Run() (/__w/llvm/llvm/build/./lib/libllvm_gtest.so.19.0git+0x40335)
#16 0x00007facd711c54d testing::internal::UnitTestImpl::RunAllTests() (/__w/llvm/llvm/build/./lib/libllvm_gtest.so.19.0git+0x5154d)
#17 0x00007facd711b889 testing::UnitTest::Run() (/__w/llvm/llvm/build/./lib/libllvm_gtest.so.19.0git+0x50889)
#18 0x00007facd730224c main (/__w/llvm/llvm/build/./lib/libllvm_gtest_main.so.19.0git+0x124c)
#19 0x00007facd6977d90 (/lib/x86_64-linux-gnu/libc.so.6+0x29d90)
#20 0x00007facd6977e40 __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x29e40)
#21 0x000000000040a905 _start (/__w/llvm/llvm/build/tools/sycl/unittests/program_manager/arg_mask/./ProgramManagerArgMaskTests+0x40a905)

--
exit: -6
--
shard JSON output does not exist: /__w/llvm/llvm/build/tools/sycl/unittests/program_manager/arg_mask/./ProgramManagerArgMaskTests-SYCL-Unit-351801-0-1.json
********************
Testing:  0.. 10.. 20.. 30.. 40.. 50.. 60.. 70.. 80.. 90.. 
********************
Failed Tests (1):
  SYCL-Unit :: program_manager/arg_mask/./ProgramManagerArgMaskTests/0/1
```

**Comment by aarongreig** [comment]
> @intel/unified-runtime-reviewers Can someone please assess the above postcommit fail and see if it could be related to the UR bump added in this PR? Thanks

we have reason to suspect that https://github.com/intel/llvm/pull/14873/ will fix this, if that merges and post-commit is still failing we will revert the bump and investigate further

**Comment by sarnex** [comment]
Seems build passes now, but there are some other postcommit failures. Will post in that PR. Thanks.

**Comment by KornevNikita** [comment]
@przemektmalon @sarnex this patch broke sycl-cts/test_event on linux L0: https://github.com/intel/llvm/actions/runs/10241858254/job/28331043363
This doesn't help - https://github.com/intel/llvm/pull/14944

Do you know possible reasons?

**Comment by sarnex** [comment]
+ @intel/bindless-images-reviewers  (see above)

Can someone please investigate the CTS issue? If we can't fix it quickly probably we should revert this patch. Thanks

**Comment by omarahmed1111** [comment]
@sarnex @KornevNikita I tried that locally and it comes from this PR: https://github.com/oneapi-src/unified-runtime/pull/1880 . I will ping the team for that to see. if there is not a quick fix, will revert that and include the reverted PR in this bump: https://github.com/intel/llvm/pull/14944

**Comment by sarnex** [comment]
Thanks Omar!

**Pull request merged**